reps/2022-04-07-serialization.md

```diff
@@ -0,0 +1,314 @@
+
+In order to resolve the above issues. We propose to refactor the current serialization code path, to
+
+1. Provide pluggable ways for users to implement custom serialization, including:
+    1. Cross-language serialization.
+    2. Out of band serialization and other optimizations.
+2. Unify the current serialization code path to this new pluggable design. Make code cleaner. Also, provide a unified interface across different languages.
+
+With a standard way to implement serializers, we can
+
+* Solve issue 4 immediately.
+* Solve issue 3 by providing build-in serializers for commonly used classes.
+* Solve issue 2 by registering different serializers for primitive types.
+* Solve issue 1 by implementing an advanced serializer with OOB optimization.
+
+### Should this change be within `ray` or outside?
+
+It should be a part of the main `ray` project.
+
+## Stewardship
+
+### Required Reviewers
+
+Hao Chen, Qing Wang, Eric Liang, Zhi Lin, Simon Mo, Jiajun Yao, Siyuan Zhuang, Clark Zinzow
+
+### Shepherd of the Proposal (should be a senior committer)
+
+Hao Chen
+
+## Design and Architecture
+
+### User-level API
+
+#### Register Serializer
+
+Firstly, if users want to implement custom serialization, they should register their serializers to Ray.
+Note that a unique string ID should be provided for class identification in cross-language serialization.
+
+```python
+# In Python
+ray.register_serializer("ArrowTable", ArrowTable, ArrowTableSerializer())
+ray.register_serializer("MyClass", MyClass, MyClassSerializer())
+```
+
+```java
+// In Java
+Ray.registerSerializer("ArrowTable", ArrowTable.class, new ArrowTableSerializer());
+Ray.registerSerializer("MyClass", MyClass.class, new MyClassSerializer());
+```
+
+Ray will maintain a map from ClassID to Serializer in memory. This relationship won't be persisted, which means you need to register them again when the process is restarted.
+
+As the first step, this API is only used by Ray core. That means we can simply hard code them without needing persistence.
+If we are to expose this API to users, we can consider to add the persistance feature in the future.
+
+We'll also provide a `ray.get_serializer` to get the serializer by either language-specific type or unique string ID, or call `ray.serialize/deserialize` to do serialization directly.
+
+```python
+ray.get_serializer("ArrowTable")
+ray.get_serializer(ArrowTable)
+
+ray.serialize(arrow_table_obj)
+ray.deserialize(ray_serialized_result)
+```
+
+#### Implement Serializer
+
+Then, the user needs to implement this custom serializer, implements the interface shown below:
+Let's take Python as an example. Other languages will be similar.
+
+```python
+class MySerializer(RaySerializer):
+    def serialize(self, object: MyClass) -> RaySerializationResult:
+        pass
+    def deserialize(self, in_band_buffer: bytes,
+                    oob_iterator: Iterator[memoryview]) -> MyClass:
```

Comment:
Consider using the same name as the field in `RaySerializationResult ` (i.e., out_of_band_buffers)?

Reply:
It's an iterator, not buffers. How about `out_of_band_buffers_iterator`?